src/toil/server/wes/toil_backend.py

```diff
@@ -0,0 +1,318 @@
+        run.set_up_run()
+
+        # stage the uploaded files to the execution directory, so that we can run the workflow file directly
+        temp_dir = run.exec_dir
+        try:
+            _, request = self.collect_attachments(run_id, temp_dir=temp_dir)
+        except ValueError:
+            run.clean_up()
+            raise
+
+        wf_type = request["workflow_type"].lower().strip()
+        version = request["workflow_type_version"]
+
+        # validate workflow request
+        supported_versions = self.supported_versions.get(wf_type, None)
+        if not supported_versions:
+            run.clean_up()
+            raise VersionNotImplementedException(wf_type)
+        if version not in supported_versions:
+            run.clean_up()
+            raise VersionNotImplementedException(wf_type, version, supported_versions)
+
+        run.set_state("QUEUED")
+
+        logger.info(f"Putting workflow {run_id} into the queue. Waiting to be picked up...")
+        run.queue_run(request, options=self.options)
+
+        return {
+            "run_id": run_id
+        }
+
+    @handle_errors
+    def get_run_log(self, run_id: str) -> Dict[str, Any]:
+        """ Get detailed info about a workflow run."""
+        run = self._get_run(run_id, should_exists=True)
+        state = run.get_state()
+
+        request = json.loads(run.fetch("request.json", "{}"))
+
+        cmd = run.fetch("cmd", "").split("\n")
+        start_time = run.fetch("start_time")
+        end_time = run.fetch("end_time")
+
+        # TODO: stdout and stderr should be a URL that points to the output file, not the actual contents.
+        stdout = run.fetch("stdout")
+        stderr = run.fetch("stderr")
+        exit_code = run.fetch("exit_code")
+
+        output_obj = {}
+        if state == "COMPLETE":
+            output_obj = run.get_output_files()
+
+        return {
+            "run_id": run_id,
+            "request": request,
+            "state": state,
+            "run_log": {
+                "cmd": cmd,
+                "start_time": start_time,
+                "end_time": end_time,
+                "stdout": stdout,
+                "stderr": stderr,
+                "exit_code": int(exit_code) if exit_code is not None else None,
+            },
+            "task_logs": [],
+            "outputs": output_obj,
+        }
+
+    @handle_errors
+    def cancel_run(self, run_id: str) -> Dict[str, str]:
+        """ Cancel a running workflow."""
+        run = self._get_run(run_id, should_exists=True)
+        state = run.get_state()
+
+        if state in ("CANCELING", "CANCELED", "COMPLETE"):
+            # We don't need to do anything.
+            logger.warning(f"A user is attempting to cancel a workflow in state: '{state}'.")
+        elif state in ("EXECUTOR_ERROR", "SYSTEM_ERROR"):
+            # Something went wrong. Let the user know.
+            raise WorkflowExecutionException(f"Workflow is in state: '{state}', which cannot be cancelled.")
```

Comment:
Should we really 500 the *cancellation* because the *workflow* didn't succeed? I wouldn't expect a 500 unless something happens that we the server devs didn't contemplate. Trying to do something that can't be done like this would really be a 400-something, like `409 Conflict`, but that's not [one of the acceptable return codes](https://github.com/ga4gh/workflow-execution-service-schemas/blob/c3b19854240c4fcbaf3483e22b19db0a918a7ee5/openapi/workflow_execution_service.swagger.yaml#L296-L315). Nor is just a plain `400`. Of the ones available, `403 Forbidden` might be the best option here, though we don't have an exception type made for that yet.